cpp/src/sort/rank.cu

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/sorting.hpp>
+#include <cudf/sorting.hpp>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/sequence.h>
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+template <bool has_nulls, typename ReturnType = bool>
+struct unique_comparator {
+  unique_comparator(table_device_view device_table,
+                    size_type const *sorted_order)
+      : comp(device_table, device_table, true), perm(sorted_order) {}
+  __device__ ReturnType operator()(size_type index) const noexcept {
+    return index == 0 || not comp(perm[index], perm[index - 1]);
+  };
+
+private:
+  row_equality_comparator<has_nulls> comp;
+  size_type const *perm;
+};
+
+std::unique_ptr<table> rank(table_view const &input, rank_method method,
```

Comment:
This is a really long function with no documentation of the algorithm. How does it work? Since there's no explanation of how it works or what it does, it's confusing that `double` and `FLOAT64` come up so often. Shouldn't rank be independent of datatype?

Reply:
rank is of `float64` type. (which is `double`). Pandas returns a `float64` type column as result.

Reply:
added some description of algorithm for each method.

Reply:
In general, I find this function incredibly hard to read due to the style / formatting. It even looks like you are using clang-format, but the 80 col limit (or whatever you are using) is just causing so much noise. Are you able to increase it? We will be doing this in the future anyway.

Reply:
I use 80. (Google format). planning to increase to 120. I hope, it won't mess up split mode of PR reviews. 
I am also waiting for repo specific `.clang-format` file PR.